Tomo - Subtomogram refinement and Initial Model

adrianq opened the pull request:
Some comments/doubt on our side:
- The transformation matrix is built simply from the 3x4 align matrix with row [0,0,0,1] appended. @cossorzano I am almost sure we are doing something wrong [here](https://github.com/scipion-em/scipion-em-eman2/pull/36/files#diff-28427d6dd9c071967e17ef6c82a16efaR476) Can you shed some light on this?
- samplingRate for outputSetOfSubTomograms should be obtained from input particles, correct? The other option is to take it from the reference...

EstrellaFG commented on eman2/protocols.conf:
```diff
@@ -41,7 +41,12 @@ Random Conical Tilt = [
 	{"tag": "section", "text": "Imports", "icon": "bookmark.png", "children": []},
 	{"tag": "section", "text": "Movies/Micrographs", "children": []},
-	{"tag": "section", "text": "Reconstruction", "children": []},
+	{"tag": "section", "text": "Reconstruction", "children": [
```
"Reconstruction" section should go after "subtomogram averaging" section, just to follow a typical tomography workflow.

adrianq replied:
Done https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487 I also move Particles Picking and Extraction to the top following the same logic

EstrellaFG commented on eman2/protocols/protocol_tomo_initialmodel.py:
```diff
@@ -0,0 +1,225 @@
+from itertools import count
+
+import numpy
+
```
Remove unused imports (I think the ones above are not used)

adrianq replied:
https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487

EstrellaFG commented on eman2/protocols/protocol_tomo_initialmodel.py:
```diff
@@ -0,0 +1,225 @@
+    """
+    This protocol wraps *e2spt_sgd.py* EMAN2 program.
+
+    It will take a set of class-averages/projections and build a set
```
I guess the protocol accepts a setOfSubTomograms in general, not especifically class-averages, but you probably know it better. However, I think they are not projections in any case, but maybe this should be confirmed

adrianq replied:
Yep, it was the wrong help. I changed the initial model and subtomogram refinement texts for some more elaborated helps here https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487

EstrellaFG commented on eman2/protocols/protocol_tomo_initialmodel.py:
```diff
@@ -0,0 +1,225 @@
+                      label='Mask',
+                      allowsNull=True,
+                      pointerClass='VolumeMask',
+                      help='Select the subtomogram to use as reference')
```
Select a mask?

adrianq replied:
https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487

EstrellaFG commented on eman2/protocols/protocol_tomo_subtomogram_refinement.py:
```diff
@@ -0,0 +1,243 @@
+from pyworkflow import utils as pwutils
+import pyworkflow.protocol.params as params
+import pyworkflow.em as pwem
+from pyworkflow.em import Transform
```
I think this import is unused

adrianq replied:
Right https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487

EstrellaFG commented on eman2/protocols/protocol_tomo_subtomogram_refinement.py:
```diff
@@ -0,0 +1,243 @@
+
+    def getOutputFile(self, folderpattern, folder, files, pattern):
+        pattern = "^" + folderpattern + pattern
+        import re
```
I think all the imports should go together at the begining

adrianq replied:
They were already at the top but forgot to delete it from here as well. Done now https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487

adrianq commented on eman2/protocols/protocol_tomo_subtomogram_refinement.py:
```diff
@@ -0,0 +1,243 @@
+        self.runJob(program, args)
+
+    def runMLStep(self, params):
+        pass
```
Old code, remove now https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/df1ed7237ecdf2e47a930f08d0b5fe8ee516a487

EstrellaFG commented:
In today's demo, eman users have said that in the initial model (outside scipion) the reference is optional, however in the scipion form of the protocol, "Reference" field can not be empty. Is it possible to do this field optional and if the user leaves the field reference empty, run eman program without any reference?

EstrellaFG commented:
@adrianq, we think that the output of initial model and refinement should be an AverageSubTomogram instead of a normal subtomogram.

adrianq commented:
@EstrellaFG I have implemented your suggestions:
- Reference is now optional (See https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/0884082e60471c8657c4b69be591faa9e220a200 and https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/1f260831788cf18c450d36a4a1f5e2954dc3a4d3). As per the eman2 documentation, it is thought to be an optional param:
> 3D reference for initial model generation. No reference is used by default.

However, I am getting an error when no reference is passed. I will keep as optional for now so the guys in the lab can give it a try and let us know if this parameter is actually required.
- Remove bold and set parameters as expert parameters (See https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/1f260831788cf18c450d36a4a1f5e2954dc3a4d3). If anybody has a better approach regarding splitting regular and expert parameter I am happy to change it
- We are now using AverageSubTomogram instead of a regular SubTomogram (See https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/001a07a6e902f3663d9f62c04f133d768e32a4e9)
- I have converted shifts from voxel to angstrom as per Carlos Oscar requests (See https://github.com/scipion-em/scipion-em-eman2/pull/36/commits/fe7f2d25cee5fc98b360fa7223f4e95a85a26177). It would be great if he can have a look at this bit of code in particular.
- As part of this last commit, I have pushed some minor fixes for the tests.

Merging this now to tomo branch. Give it a try anyway and let me know if I am missing anything! Thanks!
